Handle errors during snapshot

bhalevy opened the pull request:
This series refactors `table::snapshot` and moves the responsibility
to flush the table before taking the snapshot to the caller.

`flush_on_all` and `snapshot_on_all` helpers are added to replica::database
(by making it a peering_sharded_service) and upper layers,
including api and snapshot-ctl now call it instead of calling cf.snapshot directly.

With that, error are handed in table::snapshot and propagated
back to the callers.

Failure to allocate the `snapshot_manager` object is fatal,
similar to failure to allocate a continuation, since we can't
coordinate across the shards without it.

Test: unit(dev), rest_api(debug)

Fixes #10500

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/267/

nyh commented on api/storage_service.cc:
```diff
@@ -780,12 +780,16 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
         auto keyspace = validate_keyspace(ctx, req->param);
         auto column_family = req->get_query_param("cf");
-        return make_ready_future<json::json_return_type>(json_void());
+        auto ts = db_clock::now();
```
Wow, so this was just unimplemented? Why is this patch part of *this* series? Do we have a test for it? I would also open an issue (if we don't have one yet?) and mark this patch "Fixes" it (in case we'll want to backport it, for example).

bhalevy replied:
The patch is part of this series just to facilitate testing truncate via the api.
But the test could also issue truncate via CQL.
I'll open an issue about that (and look for other unimplemented API calls)
and we can fix it separately.

nyh replied:
Good you opened the issue, it lead to a good discussion and I think a conclusion that we don't need this REST API - your test can simply use the CQL "TRUNCATE" command, right?

bhalevy replied:
and also 
```python
@contextmanager
def new_test_snapshot(rest_api, keyspaces=[], tables=[], tag=""):
```

nyh commented on test/rest_api/test_storage_service.py:
```diff
@@ -290,7 +298,7 @@ def verify_snapshot_details(expected):
                     'key': tag1,
                     'value': [
-                        {'ks': ks0, 'cf': cf00, 'total': 1, 'live': 0},
+                        {'ks': ks0, 'cf': cf00, 'total': 0, 'live': 0},
```
I think I don't understand what this "total" means... Why is it now 0?

bhalevy replied:
`total` sums the (disk allocated) size of the sstables in the snapshot.
If the table contained no data, `total` is zero (i.e. it doesn't not include the size of the `schema.cql` and `manifest.json` files that are in the snapshot directory.

`live` sums the size of sstables that are present both in the snapshot _and_ in the table base directory (they are hard-linked together).  It can be used to estimate how much space will be freed when deleting the snapshot.

BTW, it is hard to tell by the this api alone since `total - live` is not really the unique capacity in the snapshot since the sstables that aren't considered "live" may still be hardlinked in other snapshot, so deleting a particular snapshot will not necessarily free-up all the non-live space.  In theory, we could extend the api with a `unique` counter to the snapshot details that will hold the size of sstables that have a link-count of 1 and hence are unique to this particular snapshot.

nyh commented on test/rest_api/test_storage_service.py:
```diff
@@ -277,6 +277,14 @@ def verify_snapshot_details(expected):
                 'value': [{'ks': ks0, 'cf': cf00, 'total': 1, 'live': 0}]
             })
 
+            resp = rest_api.send("POST", f"storage_service/truncate/{ks0}", { 'cf': cf00 })
```
Personally (but I guess this is just a matter of taste), I don't like tests to combine several completely unrelated features. It makes it harder to understand the test and what it really tests, and if in the future it fails, it makes it harder to understand which feature broke. I would have left the orginal test unmodified, and added a new test for a combination of snapshot and truncate - if you think there's something interesting in this combination.
But your call.

nyh replied:
Something I just thought about - this truncate/snapshot combination assumes that auto-snapshot is turned off? While test/cql-pytest/run indeed runs with auto-snapshot turns off, it was meant to be a performance improvement - not something that test correctness should rely on.

bhalevy replied:
I agree.

I have a follow up patch that adds an optional, boolean `snapshot` parameter to the truncate API
so the caller can override the default and explicitly say if a snapshot is desired or be skipped.

I should probably add a comment about the test assumption that auto-snapshot is off.

nyh replied:
> I agree.
> 
> I have a follow up patch that adds an optional, boolean `snapshot` parameter to the truncate API so the caller can override the default and explicitly say if a snapshot is desired or be skipped.

Since I think the truncate API shouldn't exist, I don't think this test should use it or improvements of it ;-)

> I should probably add a comment about the test assumption that auto-snapshot is off.

Even bette can be to *check* what is the setting of auto-snapshot, and set your expectations in the test accordingly. You can check the setting of a configuration parameter by reading from the `system.config` table - you can see an example in test/cql-pytest/conftest.py or test/cql-pytest/test_tools.py

nyh commented on replica/table.cc:
```diff
@@ -1349,30 +1349,30 @@ future<> table::snapshot(database& db, sstring name) {
-                return io_check(sync_directory, jsondir);
-            }).finally([this, &tables, &db, &jsondir] {
+        co_await io_check(sync_directory, jsondir);
+
```
I don't understand your comment about "catch immediate errors and return them". Here, if anything above here throws, the entire thing will throw, and the code below will not run. This is different from the code before your modification, which had a "finally" on the code below.

bhalevy replied:
Yes.  This is the gist of this whole series.
The existing code had a finally clause bu it didn't handle all errors.

Any error in this part: https://github.com/scylladb/scylla/blob/eeec502aee7a693ccaf84bbf4a4313350c1a26bd/replica/table.cc#L1366-L1383

would cause other shards to hang, since the semaphore wouldn't be signaled here:
https://github.com/scylladb/scylla/blob/eeec502aee7a693ccaf84bbf4a4313350c1a26bd/replica/table.cc#L1386

and here: https://github.com/scylladb/scylla/blob/eeec502aee7a693ccaf84bbf4a4313350c1a26bd/replica/table.cc#L1404

nyh replied:
What confused me was that you lost the "finally" stuff and re-added it only in the next patch - which I didn't know would happen until I moved to reviewing the next patch.

bhalevy replied:
I squashed them together in v2

nyh commented on replica/table.cc:
```diff
@@ -1384,37 +1384,25 @@ future<> table::snapshot(database& db, sstring name) {
-                    }).then([snapshot] {});
-                });
-            });
+            co_await snapshot->manifest_write.wait(1);
```
Again, this code used to be in a "finally" but no longer is. Why is this ok?

nyh commented on replica/table.cc:
```diff
@@ -1350,59 +1351,82 @@ future<> table::snapshot(database& db, sstring name) {
+            });
+        });
         co_await io_check(sync_directory, jsondir);
+    } catch (...) {
```
Maybe this answers my question on the previous patch - it seems you are adding a "catch" where we previously had finally. So perhaps we should merge these two patches?

bhalevy replied:
Maybe merging the patches would indeed make the overall change clearer and easier to review.

nyh reviewed (changes requested):
I think your series is mostly good, but I did leave a bunch of requests / comments. Thanks.

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -209,3 +209,23 @@ def insert():
             t1.join()
             t2.join()
+
+def test_storage_service_flush(cql, this_dc, rest_api):
```
Sure.  I'll add some TODO comments (since it's somewhat out of scope of this series)

bhalevy commented on replica/database.hh:
```diff
@@ -1558,6 +1558,10 @@ public:
     future<> flush(const sstring& ks, const sstring& cf);
+    future<> flush_on_all(utils::UUID id);
+    future<> flush_on_all(std::string_view ks_name, std::string_view table_name);
+    future<> flush_on_all(std::string_view ks_name, std::vector<sstring> table_names);
```
I'll add s9me documentation.

Flushing all sstables in the keyspace when no sstables are given is the api semantics, but we can check that in the api layer and call the `flush_on_all(ks)` variant instead, to make things clearer and less ambiguous.

Regarding the use of these functions: it's for flushing on all shards before taking a snapshot (hence the relevance to this series),

bhalevy replied:
I added some documentation in v2 and changed the list-accepting variant to do nothing for an empty list.
It's api caller calls the ks-only variant if no tables are provided over the api.

bhalevy commented on replica/database.cc:
```diff
@@ -2031,6 +2031,31 @@ future<> database::flush(const sstring& ksname, const sstring& cfname) {
+}
+
+future<> database::flush_on_all(std::string_view ks_name, std::string_view table_name) {
+    return flush_on_all(find_uuid(ks_name, table_name));
```
Because `find_column_family(ks, cf)` is per replica::database shard instance and returns a shard-local column_family&
the uuid returned by find_uuid(ks, cf) can be used on any shard to locate the table more efficiently
than looking up the coloumn family by (ks_name, cf_name) on every shard..

bhalevy commented on api/api-doc/storage_service.json:
```diff
@@ -617,7 +617,7 @@
                      "name":"tag",
                      "description":"the tag given to the snapshot",
-                     "required":false,
+                     "required":true,
```
I'll open an issue for that too.
It is minor (since the requirement is only at the documentation level) but might be worth backporting.

bhalevy replied:
will be considered in a separate PR.

bhalevy commented on api/api-doc/storage_service.json:
```diff
@@ -624,15 +624,15 @@
                   {
                      "name":"kn",
-                     "description":"Comma seperated keyspaces name to snapshot",
+                     "description":"Comma seperated keyspaces name to snapshot. If omitted, snapshot all keyspaces.",
```
will do.

bhalevy replied:
will be done in a separate PR.

bhalevy commented on api/api-doc/storage_service.json:
```diff
@@ -624,15 +624,15 @@
                   {
                      "name":"cf",
-                     "description":"the column family to snapshot",
+                     "description":"the table(s) to snapshot (restricted to a single keyspace). If omitted, snapshot all tables.",
```
sure

bhalevy replied:
will be considered in a separate PR.

bhalevy commented on db/snapshot-ctl.cc:
```diff
@@ -61,18 +62,21 @@ future<> snapshot_ctl::take_snapshot(sstring tag, std::vector<sstring> keyspace_
-                        return cf.snapshot(db, tag, bool(sf));
-                    });
-                });
+    return run_snapshot_modify_operation([tag = std::move(tag), keyspace_names = std::move(keyspace_names), sf, this] () mutable {
```
True, there are no functional changes in this patch.
I'll improve the commitlog.

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -229,3 +229,100 @@ def test_storage_service_flush(cql, this_dc, rest_api):
+            if data['key'] == expected['key']:
+                assert not found
+                found = True
+                sort_key = lambda v: f"{v['ks']}-{v['cf']}"
```
I want to sort a list of dictionaries and dictionaries don't have a default order in Python...
```python
>>> v = [{'cf':'mycf', 'ks': 'myks'}, {'ks': 'foo', 'cf':'bar'}]

>>> sorted(v)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'dict' and 'dict'

>>> sorted(v, key=lambda v: f"{v['ks']}-{v['cf']}")
[{'ks': 'foo', 'cf': 'bar'}, {'cf': 'mycf', 'ks': 'myks'}]
```

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -229,3 +229,100 @@ def test_storage_service_flush(cql, this_dc, rest_api):
+                    v = value[i]
+                    ev = expected_value[i]
+                    assert v['ks'] == ev['ks'], f"ks mismatch: expected {expected_value} but got {value}"
+                    assert v['cf'] == ev['cf'], f"cf mismatch: expected {expected_value} but got {value}"
```
The extra verbosity was helpful during debugging so I decided to leave it.
Without it, pytest would just report the different values at the specific assert but the context is unclear.
I prints the `expected` input, but in a concise form (with "..." if the printed output is too long)
and it doesn't print the `value` dictionary at all, so from the assert it is hard to tell what went wrong and why.

bhalevy replied:
Ah, and I can't simply compare the dicts to each other since the expected values for `total` and `live` are qualitative, not quantitative.  I don't know (and care about) what the actual size of the sstables is.
I just want to check whether `total` and `live` are either zero or non-zero in both `value` and `expected`.
Maybe if I normalize the values of `total` and `live` in the retrieved `value`, I could do a trivial dict comparison.

nyh replied:
> The extra verbosity was helpful during debugging so I decided to leave it. Without it, pytest would just report the different values at the specific assert but the context is unclear. I prints the `expected` input, but in a concise form (with "..." if the printed output is too long) and it doesn't print the `value` dictionary at all, so from the assert it is hard to tell what went wrong and why.

In my experience, it always prints the arrays being compared and the specific elements that were found different, even if sometimes you need to think a bit about what these elements mean.

Anyway, I won't insist here - I'll "resolve" this discussion. This was more of a general "warning" to people who switch to pytest from other systems like our old nosetest-based dtest or boost tests in C++, and got used to write all these elaborate comparison functions, while most of them aren't necessary in pytest (as you said, some normalization functions do come in handy, though, but those are usually much less verbose than comparison function, and you usually don't end up with a gazillion of them).

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -229,3 +229,100 @@ def test_storage_service_flush(cql, this_dc, rest_api):
+                        {'ks': ks0, 'cf': cf00, 'total': 1, 'live': 0},
+                        {'ks': ks0, 'cf': cf01, 'total': 1, 'live': 0}
+                    ]
+                })
```
yes.

And maybe it would be best to implement it with the `contextmanager`.
I'll look into it.

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -229,3 +229,100 @@ def test_storage_service_flush(cql, this_dc, rest_api):
+                            'value': [
+                                {'ks': ks0, 'cf': cf00, 'total': 1, 'live': 0},
+                                {'ks': ks0, 'cf': cf01, 'total': 1, 'live': 0},
+                                {'ks': ks1, 'cf': cf10, 'total': 0, 'live': 0}
```
See answer above.  Basically total is 0 when taking a snapshot of an empty table.

nyh replied:
I see now. I don't like this API (who can guess that "total" means total size and not total number? what units does this size even have? where is this documented?), but obviously this is a test for the existing API, and not the time to change it.

bhalevy replied:
unit is bytes.  FWIW there is some documentation here: https://cassandra.apache.org/doc/trunk/cassandra/tools/nodetool/listsnapshots.html
and I agree that it's out of scope of this PR... :)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/280/

nyh commented on test/rest_api/rest_util.py:
```diff
@@ -0,0 +1,32 @@
+#   with new_test_snapshot(cql, tag, keyspace, [table(s)]) as snapshot:
+# This is not a fixture - see those in conftest.py.
+@contextmanager
+def new_test_snapshot(rest_api, keyspaces=[], tables=[], tag=""):
```
In my opinion (but I won't hold the series because of this), since this is a tool specifically for **snapshots**, it belongs in the only file that tests snapshots (test_storage_service.py) and we don't need a separate source file for it. I'm not a fan of used-once "utilities". But as I said, I won't insist on this.

nyh replied:
By the way, why does tag default to ""? Is this even legal?

bhalevy replied:
It defaults to `f"test_snapshot_{int(time.time() * 1000)}"` following the convention used elsewhere.
`new_test_snapshot` yields it as the result so it makes sense to leave it for the function if not provided by the user.

nyh commented on db/snapshot-ctl.cc:
```diff
@@ -104,10 +104,8 @@ future<> snapshot_ctl::do_take_column_family_snapshot(sstring ks_name, std::vect
-            if (table_name.find(".") != sstring::npos) {
-                throw std::invalid_argument("Cannot take a snapshot of a secondary index by itself. Run snapshot on the table that owns the index.");
+            if (cf.schema()->is_view()) {
+                throw std::invalid_argument("Do not take a snapshot of a materialized view or a secondary index by itself. Run snapshot on the base table instead.");
```
A question (as is often the case, the best way to answer such a question is with a test ;-)) - when you take a snapshot of a table with views, does it actually snapshot the views as well?

bhalevy replied:
No, it doesn't.  The idea, as far as I understand it, is that one can always re-generate MV/SI from the snapshot of the base table. Also, it is quite complicated to take a _coherent_ snapshot of both the base table and its view so one could restore both and hope they'll be in sync - hence it's better to regenerate MV/SI from the base table if needed rather than to go through the extra hoop of doing a distributed, coherent snapshot of both the base table and its view.

nyh commented:
@bhalevy I see you forced-push the branch yet again, less than a minute before my merge. My merge probably didn't include that latest change - but please verify.

bhalevy commented:
In v2 (1d39d803affd3c6e49d2319ef1591d2d79ac188f):
- `test_storage_service_flush`:
  - added negative test cases for simple error scenarios and added TODO comment
  - simplified verification (by normalizing `total` and `live` and comparing whole `value` vs. `expected` dicts
  - lamda function for sorting is still required...
- `flush_on_all`:
  - added brief documentation in header file
  - empty sstables vector is noop, caller should call the keyspace-only variant to flush all tables in a keyspace
  - (same principle applied for snapshot_on_all in a later patch)
- dropped implementation of truncate api
  - added `new_test_snapshot` `@contextmanager` function in a new `rest_api/rest_util.py` file
- "api: storage_service: take_snapshot: improve api help messages": rewrote description, fixing spelling errors and typos and clarifying what the api call does.
- Added the following patches:
  - "api: storage_service: coroutinize take_snapshot and del_snapshot"
  - "api: storage_service: increase visibility of snapshot ops in the log"
- "snapshot-ctl: refactor and coroutinize take_snapshot / take_column_family_snapshot": clarified there's no functional change in commitlog
- "snapshot-ctl: run_snapshot_modify_operation: reject views and secondary index using the schema"
  - new patch
  - Fixes #10526
  - comes with an additional rest_api unit test
    - uses new_materialized_view
    - adds and uses new_secondary_index in `cql-pytest/util.py`
- No changes in:
  - database: add snapshot_on_all
  - table: make snapshot method private
  - database: snapshot_on_all: flush before snapshot if needed
- "test: rest_api: storage_service: verify_snapshot_details: add truncate"
  - use CQL TRUNCATE rather than truncate api
- No changes in:
  - database: truncate: skip flush when taking snapshot
  - table: snapshot: get rid of skip_flush param
- table: snapshot: handle errors
  - squashed together with coroutinization patch that was separate in v1
  - otherwise, no functional changes.

bhalevy commented:
> @bhalevy I see you forced-push the branch yet again, less than a minute before my merge. My merge probably didn't include that latest change - but please verify.

Thanks @nyh.  You merged the latest version (v2, 1d39d803affd3c6e49d2319ef1591d2d79ac188f)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/284/
